builtin/credential/approle/path_role.go

```diff
@@ -0,0 +1,1544 @@
+	}
+
+	if role, err := b.roleEntry(req.Storage, strings.ToLower(roleName)); err != nil {
+		return nil, err
+	} else if role == nil {
+		return nil, nil
+	} else {
+		// Convert the 'time.Duration' values to second.
+		role.SecretIDTTL /= time.Second
+		role.TokenTTL /= time.Second
+		role.TokenMaxTTL /= time.Second
+		role.Period /= time.Second
+
+		// Create a map of data to be returned and remove sensitive information from it
+		data := structs.New(role).Map()
+		delete(data, "role_id")
+		delete(data, "hmac_key")
+
+		return &logical.Response{
+			Data: data,
+		}, nil
+	}
+}
+
+// pathRoleDelete removes the role from the storage
+func (b *backend) pathRoleDelete(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleName := data.Get("role_name").(string)
+	if roleName == "" {
+		return logical.ErrorResponse("missing role_name"), nil
+	}
+
+	role, err := b.roleEntry(req.Storage, strings.ToLower(roleName))
+	if err != nil {
+		return nil, err
+	}
+
+	// Acquire the lock before deleting the secrets.
+	b.roleLock.Lock()
+	defer b.roleLock.Unlock()
+
+	// Just before the role is deleted, remove all the SecretIDs issued as part of the role.
+	if err = b.flushRoleSecrets(req.Storage, role.RoleID); err != nil {
+		return nil, fmt.Errorf("failed to invalidate the secrets belonging to role '%s': %s", roleName, err)
+	}
+
+	// Delete the reverse mroleing from RoleID to the role
+	if err = b.roleIDEntryDelete(req.Storage, role.RoleID); err != nil {
+		return nil, fmt.Errorf("failed to delete the mroleing from RoleID to role '%s': %s", roleName, err)
+	}
+
+	// After deleting the SecretIDs and the RoleID, delete the role itself
+	if err = req.Storage.Delete("role/" + strings.ToLower(roleName)); err != nil {
+		return nil, err
+	}
+
+	return nil, nil
+}
+
+// Returns the properties of the SecretID
+func (b *backend) pathRoleSecretIDAccessorRead(req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+	roleName := data.Get("role_name").(string)
+	if roleName == "" {
+		return logical.ErrorResponse("missing role_name"), nil
+	}
+
+	secretIDAccessor := data.Get("secret_id_accessor").(string)
+	if secretIDAccessor == "" {
+		return logical.ErrorResponse("missing secret_id_accessor"), nil
+	}
+
+	// SecretID is indexed based on salted RoleID and HMACed SecretID.
+	// Get the role details to fetch the RoleID and accessor to get
+	// the HMAC-ed SecretID.
+
+	role, err := b.roleEntry(req.Storage, strings.ToLower(roleName))
+	if err != nil {
+		return nil, err
+	}
+	if role == nil {
+		return nil, fmt.Errorf("role %s does not exist", roleName)
+	}
+
+	accessorEntry, err := b.secretIDAccessorEntry(req.Storage, secretIDAccessor)
+	if err != nil {
+		return nil, err
+	}
+	if accessorEntry == nil {
+		return nil, fmt.Errorf("failed to find accessor entry for secret_id_accessor:%s\n", secretIDAccessor)
+	}
+
+	entryIndex := fmt.Sprintf("secret_id/%s/%s", b.salt.SaltID(role.RoleID), accessorEntry.SecretIDHMAC)
```

Comment:
Reminder to switch these to `secret_id/<HMAC(role name, role HMAC key)>/secretIDHMAC`